imap_processing/idex/idex_packet_parser.py

```diff
@@ -0,0 +1,286 @@
+
+        Returns:
+            Populates {object}.data, where data is an xarray DataSet
+
+        Notes:
+            Currently assumes one L0 file will generate exactly one l1a file
+
+            There are a few things that must be determined in the future:
+            * high_sample_rate and low_sample_rate must be multiplied by
+              some scale factor that will be decided upon in the future
+            * IDEX_Trigger is a placeholder variable for now
+        """
+
+        idex_xtce = f"{packet_definition_directory}/idex_packet_definition.xml"
+        idex_definition = xtcedef.XtcePacketDefinition(xtce_document=idex_xtce)
+        idex_parser = parser.PacketParser(idex_definition)
+
+        idex_binary_data = bitstring.ConstBitStream(filename=idex_packet_file)
+        idex_packet_generator = idex_parser.generator(idex_binary_data)
+
+        self.coords = {}
+        self.data = {}
+        self.dust_events = {}
+        self.scitype_to_names = {
+            2: "TOF_High",
+            4: "TOF_Low",
+            8: "TOF_Mid",
+            16: "Target_Low",
+            32: "Target_High",
+            64: "Ion_Grid",
+        }
+
+        for packet in idex_packet_generator:
+            if "IDX__SCI0TYPE" in packet.data:
+                scitype = packet.data["IDX__SCI0TYPE"].raw_value
+                event_number = packet.data["IDX__SCI0EVTNUM"].derived_value
+                if scitype == 1:
+                    time_of_impact = (
+                        packet.data["SHCOARSE"].derived_value
+                        + TWENTY_MICROSECONDS * packet.data["SHFINE"].derived_value
+                    )
+                    self._log_packet_info(
+                        time_of_impact, packet
+                    )  # These are for our logs
+                    self.dust_events[event_number] = IDEXRawDustEvent(time_of_impact)
+                if (
+                    scitype in self.scitype_to_names
+                ):  # Populate the IDEXRawDustEvent with 1's and 0's
+                    raw_science_data = packet.data["IDX__SCI0RAW"].raw_value
+                    self.dust_events[event_number].append_raw_data(
+                        scitype, raw_science_data
+                    )
+
+        # Parse the waveforms according to the scitype present
+        # (high gain and low gain channels encode waveform data differently).
+        processed_dust_impact_list = []
+        for event_number in self.dust_events:
+            processed_dust_impact_list.append(self.dust_events[event_number].process())
+
+        self.data = xr.concat(processed_dust_impact_list, dim="Epoch")
+
+    def _log_packet_info(self, time_of_impact, pkt):
+        """
+        This function exists solely to log the parameters in the L0
+        packet, nothing here should affect the data
+        """
+        event_num = pkt.data["IDX__SCI0EVTNUM"].derived_value
+        logging.info(f"^*****Event header {event_num}******^")
```

Comment:
Do we always want these logs, or do we only want them while debugging?

`logging.debug()`